packages/openapi-v3/src/generate-schema.ts

```diff
@@ -17,26 +20,31 @@ import {SchemaObject} from './types';
 export function resolveSchema(
   fn?: Function,
   schema: SchemaObject = {},
+  options: SchemaOptions = {},
 ): SchemaObject {
   let resolvedSchema: SchemaObject = {};
 
   if (typeof fn === 'function') {
+    const noTsType = !(options && options[TS_TYPE_KEY]);
     if (fn === String) {
-      resolvedSchema = {type: 'string'};
+      resolvedSchema = { type: 'string' };
     } else if (fn === Number) {
-      resolvedSchema = {type: 'number'};
+      resolvedSchema = { type: 'number' };
     } else if (fn === Boolean) {
-      resolvedSchema = {type: 'boolean'};
+      resolvedSchema = { type: 'boolean' };
     } else if (fn === Date) {
-      resolvedSchema = {type: 'string', format: 'date-time'};
-    } else if (fn === Object) {
-      resolvedSchema = {type: 'object'};
+      resolvedSchema = { type: 'string', format: 'date-time' };
+    } else if (fn === Object && noTsType) {
+      resolvedSchema = { type: 'object' };
     } else if (fn === Array) {
-      resolvedSchema = {type: 'array'};
+      resolvedSchema = { type: 'array' };
     } else {
-      resolvedSchema = {$ref: `#/components/schemas/${fn.name}`};
+      // schemaName will override the function name
+      const schemaName = options && options.schemaName;
+      const titleFromMeta = fn.name + getTitleSuffix(options);
+      resolvedSchema = { $ref: `#/components/schemas/${schemaName || titleFromMeta}` };
     }
   }
 
-  return Object.assign(schema, resolvedSchema);
+  return Object.assign(schema, resolvedSchema, { options: Object.assign(options, { isVisited: true }) });
```

Comment:
I find it suspicious that you are setting `isVisited: true` here in a function that's not actually reading `isVisited` at all. Please move that change to the function that's reading `isVisited` value.

Reply:
that flag is not needed any more in the new PoC https://github.com/strongloop/loopback-next/pull/3466